File: android/app/src/main/java/com/zillit/drive/domain/model/DriveModels.kt
package com.zillit.drive.domain.model

data class DrivePermissions(
    val canView: Boolean = true,
    val canEdit: Boolean = false,
    val canDownload: Boolean = false,
    val canDelete: Boolean = false
)

enum class DriveSection(val apiValue: String, val displayName: String) {
    MY_DRIVE("mine", "My Drive"),
    SHARED_WITH_ME("shared", "Shared with me")
}

data class DriveFile(
    val id: String,
    val fileName: String,
    val fileExtension: String,
    val fileSizeBytes: Long,
    val mimeType: String,
    val folderId: String?,
    val filePath: String?,
    val description: String?,
    val createdBy: String,
    val createdOn: Long,
    val updatedOn: Long,
    val deletedOn: Long = 0,
    val isFavorite: Boolean = false,
    val thumbnailUrl: String? = null,
    val userPermissions: DrivePermissions? = null
)

data class DriveFolder(
    val id: String,
    val folderName: String,
    val parentFolderId: String?,
    val description: String?,
    val createdBy: String,
    val createdOn: Long,
    val updatedOn: Long,
    val deletedOn: Long = 0,
    val isFavorite: Boolean = false,
    val fileCount: Int = 0,
    val folderCount: Int = 0,
    val userPermissions: DrivePermissions? = null
)

data class DriveContents(
    val folders: List<DriveFolder>,
    val files: List<DriveFile>,
    val totalFolders: Int = 0,
    val totalFiles: Int = 0
)

data class DriveTag(
    val id: String,
    val name: String,
    val color: String,
    val projectId: String
)

data class DriveComment(
    val id: String,
    val fileId: String,
    val userId: String,
    val text: String,
    val createdOn: Long,
    val updatedOn: Long
)

data class DriveVersion(
    val id: String,
    val fileId: String,
    val versionNumber: Int,
    val fileSizeBytes: Long,
    val filePath: String,
    val createdBy: String,
    val createdOn: Long
)

data class DriveActivity(
    val id: String,
    val action: String,
    val itemId: String,
    val itemType: String,
    val userId: String,
    val details: String?,
    val createdOn: Long
)

data class FolderAccess(
    val userId: String,
    val folderId: String,
    val role: String, // owner, editor, viewer
    val inherited: Boolean = false
)

data class FileAccess(
    val userId: String,
    val fileId: String,
    val canView: Boolean,
    val canEdit: Boolean,
    val canDownload: Boolean
)

data class StorageUsage(
    val usedBytes: Long,
    val totalBytes: Long,
    val fileCount: Int
)

data class UploadSession(
    val uploadId: String,
    val s3UploadId: String,
    val fileName: String,
    val fileSizeBytes: Long,
    val chunkSize: Long,
    val totalParts: Int,
    val presignedUrls: List<PresignedUrl>,
    val folderId: String?
)

data class PresignedUrl(
    val partNumber: Int,
    val url: String
)

data class UploadPart(
    val partNumber: Int,
    val etag: String
)

data class ShareLink(
    val url: String,
    val expiresAt: String?
)

sealed class DriveItem {
    data class File(val file: DriveFile) : DriveItem()
    data class Folder(val folder: DriveFolder) : DriveItem()

    val id: String
        get() = when (this) {
            is File -> file.id
            is Folder -> folder.id
        }

    val name: String
        get() = when (this) {
            is File -> file.fileName
            is Folder -> folder.folderName
        }

    val createdOn: Long
        get() = when (this) {
            is File -> file.createdOn
            is Folder -> folder.createdOn
        }

    val isFavorite: Boolean
        get() = when (this) {
            is File -> file.isFavorite
            is Folder -> folder.isFavorite
        }

    val createdBy: String
        get() = when (this) {
            is File -> file.createdBy
            is Folder -> folder.createdBy
        }

    val userPermissions: DrivePermissions?
        get() = when (this) {
            is File -> file.userPermissions
            is Folder -> folder.userPermissions
        }
}

File: ios/ZillitDrive/Domain/Models/DriveModels.swift
import Foundation

struct DrivePermissions: Equatable {
    let canView: Bool
    let canEdit: Bool
    let canDownload: Bool
    let canDelete: Bool
}

enum DriveSection: String, CaseIterable {
    case myDrive
    case sharedWithMe

    var displayName: String {
        switch self {
        case .myDrive: return "My Drive"
        case .sharedWithMe: return "Shared with me"
        }
    }

    var quickFilter: String {
        switch self {
        case .myDrive: return "mine"
        case .sharedWithMe: return "shared"
        }
    }
}

struct DriveFile: Identifiable, Equatable {
    let id: String
    let fileName: String
    let fileExtension: String
    let fileSizeBytes: Int64
    let mimeType: String
    let folderId: String?
    let filePath: String?
    let description: String?
    let createdBy: String
    let createdOn: Int64
    let updatedOn: Int64
    let thumbnailUrl: String?
    var isFavorite: Bool = false
    var userPermissions: DrivePermissions? = nil
}

struct DriveFolder: Identifiable, Equatable {
    let id: String
    let folderName: String
    let parentFolderId: String?
    let description: String?
    let createdBy: String
    let createdOn: Int64
    let updatedOn: Int64
    let fileCount: Int
    let folderCount: Int
    var isFavorite: Bool = false
    var userPermissions: DrivePermissions? = nil
}

struct DriveContents {
    let folders: [DriveFolder]
    let files: [DriveFile]
    let totalFolders: Int
    let totalFiles: Int
}

enum DriveItem: Identifiable, Equatable {
    case file(DriveFile)
    case folder(DriveFolder)

    var id: String {
        switch self {
        case .file(let f): return f.id
        case .folder(let f): return f.id
        }
    }

    var name: String {
        switch self {
        case .file(let f): return f.fileName
        case .folder(let f): return f.folderName
        }
    }

    var isFavorite: Bool {
        switch self {
        case .file(let f): return f.isFavorite
        case .folder(let f): return f.isFavorite
        }
    }

    var createdOn: Int64 {
        switch self {
        case .file(let f): return f.createdOn
        case .folder(let f): return f.createdOn
        }
    }

    var createdBy: String {
        switch self {
        case .file(let f): return f.createdBy
        case .folder(let f): return f.createdBy
        }
    }

    var userPermissions: DrivePermissions? {
        switch self {
        case .file(let f): return f.userPermissions
        case .folder(let f): return f.userPermissions
        }
    }
}

struct DriveTag: Identifiable, Equatable {
    let id: String
    let name: String
    let color: String
}

struct DriveComment: Identifiable, Equatable {
    let id: String
    let fileId: String
    let userId: String
    let text: String
    let createdOn: Int64
}

struct DriveVersion: Identifiable, Equatable {
    let id: String
    let fileId: String
    let versionNumber: Int
    let fileSizeBytes: Int64
    let createdBy: String
    let createdOn: Int64
}

struct DriveActivity: Identifiable, Equatable {
    let id: String
    let action: String
    let itemId: String
    let itemType: String
    let userId: String
    let details: String?
    let createdOn: Int64
}

struct StorageUsage {
    let usedBytes: Int64
    let totalBytes: Int64
    let fileCount: Int
}

struct ShareLink {
    let url: String
    let expiresAt: String?
}

struct BreadcrumbItem: Identifiable, Equatable {
    let id: String?  // nil for root
    let name: String
}
